mender-image-prepare.sh

```diff
@@ -0,0 +1,305 @@
+
+  new_size_sectors=$(( $new_size / 512 ))
+
+  echo -e "Root filesystem size:"
+  echo -e "\nminimal: $(( $min_size_blocks * $block_size ))"
+  echo -e "\naligned: $new_size"
+  echo -e "\nsectors: $new_size_sectors"
+
+  sed -e 's/\s*\([\+0-9a-zA-Z]*\).*/\1/' << EOF | fdisk /dev/loop0
+	d
+	n
+	p
+	1
+	8192
+	+${new_size_sectors}
+	a
+	w # write the partition table
+	q # and we're done
+EOF
+
+  partprobe
+  endsector=($(fdisk -l -u /dev/loop0 | grep '^/dev' | tr -s ' ' | cut -d' ' -f4))
+  losetup -d /dev/loop0
+  echo "Endsector: $endsector"
+  truncate -s $((($endsector+1)*512)) $IMAGE
+  prootfs_size=$new_size_sectors
+}
+
+map_image() {
+  loopline=$(kpartx -v -a $IMAGE | grep 'loop' | cut -d' ' -f3)
+  LOOPMAP=${loopline}
+  LOOPDEV=${loopline%p*}
+  echo "Dev: $LOOPDEV, mapper: $LOOPMAP"
+
+  mkdir -p $BASE_DIR/mnt/rootfs
+  mount /dev/mapper/"${LOOPMAP}" $BASE_DIR/mnt/rootfs
+}
+
+mount_target_dev() {
+  partitions=($(fdisk -l ${TGTDEV} | grep '^/dev' | cut -d' ' -f1))
+
+  number_of_parts=${#partitions[@]}
+
+  [[ $number_of_parts -eq 4 ]] || { echo "Error: incorect device partitioning."; exit 1; }
+
+  echo "Mount target device's $number_of_parts partitions..."
+
+  for partition in "${partitions[@]}"
+  do
+    part_no=$(get_part_number_from_device $partition)
+    path=$BASE_DIR/mnt/${part_names[${part_no} - 1]}
+    mkdir -p ${path}
+
+    mount $partition $path
+  done 
+}
+
+set_uenv() {
+  [ ! -d "./mnt/boot" ] && { echo "Error: check mounted boot location."; exit 1; }
+  cat <<- 'EOF' > mnt/boot/uEnv.txt
+	loadaddr=0x82000000
+	fdtaddr=0x88000000
+	rdaddr=0x88080000
+
+	initrd_high=0xffffffff
+	fdt_high=0xffffffff
+
+	loadximage=echo debug: [/boot/vmlinuz-${uname_r}] ... ; load mmc 0:2 ${loadaddr} /boot/vmlinuz-${uname_r}
+	loadxfdt=echo debug: [/boot/dtbs/${uname_r}/${fdtfile}] ... ;load mmc 0:2 ${fdtaddr} /boot/dtbs/${uname_r}/${fdtfile}
+	loadxrd=echo debug: [/boot/initrd.img-${uname_r}] ... ; load mmc 0:2 ${rdaddr} /boot/initrd.img-${uname_r}; setenv rdsize ${filesize}
+	loaduEnvtxt=load mmc 0:2 ${loadaddr} /boot/uEnv.txt ; env import -t ${loadaddr} ${filesize};
+	check_dtb=if test -n ${dtb}; then setenv fdtfile ${dtb};fi;
+	loadall=run loaduEnvtxt; run check_dtb; run loadximage; run loadxrd; run loadxfdt;
+
+	mmcargs=setenv bootargs console=tty0 console=${console} ${optargs} ${cape_disable} ${cape_enable} root=/dev/mmcblk0p2 rootfstype=${mmcrootfstype} ${cmdline}
+
+	uenvcmd=run loadall; run mmcargs; echo debug: [${bootargs}] ... ; echo debug: [bootz ${loadaddr} ${rdaddr}:${rdsize} ${fdtaddr}] ... ; bootz ${loadaddr} ${rdaddr}:${rdsize} ${fdtaddr};
+	EOF
+}
+
+set_bootloader() {
+  echo "Setting bootloader..."
+  [ ! -d "$BASE_DIR/mnt/rootfs/opt/backup/uboot" ] && { echo "Error: check mounted location."; exit 1; }
+
+  cd $BASE_DIR/mnt/rootfs/opt/backup/uboot
+  [[ ! -f MLO || ! -f u-boot.img ]] && { echo "Error: cannot find bootloader."; exit 1; } 
+
+  cp MLO $BASE_DIR/mnt/boot/
+  cp u-boot.img $BASE_DIR/mnt/boot/
```

Comment:
This seems to be very board dependent.  We will probably need a way to conditionalize this on board type.

Reply:
Indeed. See also my comment about GRUB above.

Reply:
True, I will modify it when the whole path for Raspbian is added.

Reply:
I would suggest to have an argument that takes X number of files that will be copied to `/boot`.

Also we need to make sure that we extract the original content of the boot part of the original image. There could be other files there, e.g. raspberrypi has a lot of configuration and firmware files there along side the u-boot.bin